### PR TITLE
Cleanup sendsketch to prokka

### PR DESCRIPTION
**Pull request by boulund**
I just created this PR so we can more easily overview what is left to fix before merging to develop.

@emilio-r and I had a quick look at the sendsketch to prokka script this morning.

### CODE REVIEW

**Review by abhi18av — APPROVED** [review]
The python code needed some love for sure! I think we can go ahead with the merge, I trust, this has been tested locally.

**Comment by boulund** [comment]
I'm not sure about the status of the testing of this. @emilio-r is this ready to be merged?

**Comment by emilio-r** [comment]
While I have not been able to test the pipe (HPC still down), I can confirm that the script now works and if @abhi18av have confirmed that it runs locally then yes @boulund, it should be ready to merge.

**Comment by boulund** [comment]
I got the impression that @abhi18av was asking for confirmation that we've tested it locally, not that he tested it himself :)

If you're convinced that the script works @emilio-r , I think we can merge to develop.

**Comment by emilio-r** [comment]
Ah, yes that makes more sense. Yes, the script does work as intended when tested outside of the pipe. Now I see that we can again submit jobs to the HPC so I will try and test this (or the merged develop) sometime later today.

**Review by thorellk — APPROVED** [review]

**Comment by abhi18av** [comment]
@emilio-r please feel free to merge this once you've confirmed it on the HPC :)

**Comment by emilio-r** [comment]
I have tried to run the branch on the HPC now but run into this error, which have nothing to do with the script:

N E X T F L O W  ~  version 21.01.1-edge
Launching `/proj/uppstore2017270/Emilio/bactpipe-testing/tests/210119/BACTpipe/bactpipe.nf` [pedantic_raman] - revision: a8340c3f3b
Downloading plugin nf-amazon@1.0.1
Failed to write file '/home/emilir/.nextflow/plugins/nf-amazon-1.0.1' to plugins folder

And from .nextflow.log:

Jan-19 08:54:10.467 [main] ERROR nextflow.cli.Launcher - @unknown
org.pf4j.PluginRuntimeException: Failed to write file '/home/emilir/.nextflow/plugins/nf-amazon-1.0.1' to plugins folder
	at nextflow.plugin.PluginUpdater.download0(PluginUpdater.groovy:170)
	at nextflow.plugin.PluginUpdater.access$0(PluginUpdater.groovy)
	at nextflow.plugin.PluginUpdater$_safeDownload_closure2.doCall(PluginUpdater.groovy:188)
	at nextflow.plugin.PluginUpdater$_safeDownload_closure2.call(PluginUpdater.groovy)
	at nextflow.file.FileMutex.lock(FileMutex.groovy:107)
	at nextflow.plugin.PluginUpdater.safeDownload(PluginUpdater.groovy:188)
	at nextflow.plugin.PluginUpdater.load0(PluginUpdater.groovy:217)
	at nextflow.plugin.PluginUpdater.installPlugin(PluginUpdater.groovy:140)
	at nextflow.plugin.PluginUpdater.prepareAndStart(PluginUpdater.groovy:88)
	at nextflow.plugin.PluginsFacade.start(PluginsFacade.groovy:194)
	at nextflow.plugin.PluginsFacade.start(PluginsFacade.groovy:199)
	at nextflow.plugin.PluginsFacade.setup(PluginsFacade.groovy:161)
	at nextflow.plugin.Plugins.setup(Plugins.groovy:40)
	at nextflow.cli.CmdRun.run(CmdRun.groovy:268)
	at nextflow.cli.Launcher.run(Launcher.groovy:475)
	at nextflow.cli.Launcher.main(Launcher.groovy:657)
Caused by: java.nio.file.AtomicMoveNotSupportedException: /scratch/17938909/pf4j-update-downloader5690633583307575560/nf-amazon-1.0.1 -> /home/emilir/.nextflow/plugins/nf-amazon-1.0.1: Invalid cross-device link
	at sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:394)
	at sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:262)
	at java.nio.file.Files.move(Files.java:1395)
	at nextflow.plugin.PluginUpdater.download0(PluginUpdater.groovy:167)
	... 15 common frames omitted

As I understand this stems from the file (plugin) not being able to be copied (symlinked?) from the temporary storage (scratch) to the local storage. Is this something other HPC users can confirm or anyone offer any insight into?

**Comment by thorellk** [comment]
I am testing it now (master branch though) and it seems like it runs fine, at least it is halfway through shovill.

**Comment by emilio-r** [comment]
I downgraded to the nextflow version that @thorellk is running (21.01.0-edge) and now it seems to work again as intended. I will see if it gets through my tests and then merge, if nothing else appears.

**Comment by emilio-r** [comment]
Well, it seems that it was my turn to run into this old thing:

Adding /crex/proj/uppstore2017270/Emilio/bactpipe-testing/tests/210119/70eaeb1-downgrade/work/conda/env-3b12f71134feb842065aa76592fe3f68/opt/bbmap-38.87-0/resources/blacklist_refseq_merged.sketch to blacklist.
0.064 seconds.
Loaded 1 sketch in 0.115 seconds.
Total Time: 	2.044 seconds.
Traceback (most recent call last):
  File "/proj/uppstore2017270/Emilio/bactpipe-testing/tests/210119/BACTpipe/bin/sendsketch_to_prokka.py", line 85, in <module>
    output_species, genus, output_stain = parse_sendsketch(args.sketch, stain_db)
  File "/proj/uppstore2017270/Emilio/bactpipe-testing/tests/210119/BACTpipe/bin/sendsketch_to_prokka.py", line 65, in parse_sendsketch
    genus_two = genus_two_line.split("\t")[11].split(" ")[0]
IndexError: list index out of range

The failed sample should have passed as a clean Helicobacter pylori (and does have a genus two) and has done so multiple times in previous test runs. Apart from this one sample, all others passed through the entire pipe and did so with the correct output. So, should we still merge this to develop since this is an old issue?

File: bin/sendsketch_to_prokka.py [hunk 1]
```diff
@@ -1,35 +1,90 @@
...
+
+        if not sketch_lines[2].startswith("WKID\t"):
+            print("ERROR: sketch file '{}' does not appear to contain valid sendsketch.sh output".format(args.sketch))
+            exit(1)
+
+        genus_one_line = sketch_lines[3]
+        try:
+            genus_one = genus_one_line.split("\t")[11].split(" ")[0]
+            species_one = genus_one_line.split("\t")[11].split(" ")[1]
+        except ValueError as e:
```
Comment (boulund) [inline bin/sendsketch_to_prokka.py:59]: This should except IndexError, not ValueError, my bad!

```suggestion
        except IndexError as e:
```

File: bin/sendsketch_to_prokka.py [hunk 2]
```diff
@@ -1,35 +1,90 @@
...
+            print("ERROR: Could not parse line 4 of '{}':\n{}".format(args.sketch, genus_one_line))
+            exit(1)
+
+        try:
+            genus_two_line = sketch_lines[4]
+            genus_two = genus_two_line.split("\t")[11].split(" ")[0]
+        except ValueError as e:
```
Comment (boulund) [inline bin/sendsketch_to_prokka.py:66]: This should except IndexError, not ValueError, my bad!

```suggestion
        except IndexError as e:
```

**Comment by boulund** [comment]
Great that you ran into this @emilio-r ! We should have intercepted `IndexError` and not `ValueError`. If we fix that all should be good.

**Comment by emilio-r** [comment]
Ok great, thank you @boulund for that tip. However, changing that (_and only that_) now does this:

         BACTpipe workflow completed without errors         
               Check output files in folder:                
                      BACTpipe_results                      
============================================================

executor >  slurm (50)
[e8/97a0b2] process > FASTP (CAN-152_S19_L001)       [100%] 10 of 10 ✔
[3f/ccf875] process > SHOVILL (CAN-152_S19_L001)     [100%] 10 of 10 ✔
[87/7ad566] process > ASSEMBLY_STATS (CAN-152_S19... [100%] 10 of 10 ✔
[28/1a29ee] process > SCREEN_FOR_CONTAMINANTS (CA... [100%] 10 of 10, failed:...
[22/403461] process > PROKKA (CAN-152_S19_L001)      [100%] 9 of 9 ✔
[b5/d6039d] process > MULTIQC                        [100%] 1 of 1 ✔

From .command.log of that failed SCREEN:

nxf-scratch-dir r168:/scratch/17940541/nxf.miWht3YR8P
Adding /crex/proj/uppstore2017270/Emilio/bactpipe-testing/tests/210119/70eaeb1-indexerror/work/conda/env-3b12f71134feb842065aa76592fe3f68/opt/bbmap-38.87-0/resources/blacklist_refseq_merged.sketch to blacklist.
0.061 seconds.
Exception in thread "main" java.lang.AssertionError: BBSketch does not currently support empty files.
0, 0, 0, 10000, true
	at sketch.SketchObject.toSketchSize(SketchObject.java:911)
	at sketch.SketchHeap.maxLen(SketchHeap.java:152)
	at sketch.SketchHeap.toSketchArray_minCount(SketchHeap.java:161)
	at sketch.Sketch.<init>(Sketch.java:56)
	at sketch.Sketch.<init>(Sketch.java:52)
	at sketch.SketchMakerMini.processInner(SketchMakerMini.java:175)
	at sketch.SketchMakerMini.toSketches(SketchMakerMini.java:116)
	at sketch.SketchTool.loadSketchesFromSequenceFile(SketchTool.java:376)
	at sketch.SketchTool.loadSketchesFromFile(SketchTool.java:343)
	at sketch.SketchTool.loadSketches_MT(SketchTool.java:291)
	at sketch.SketchTool.loadSketches_MT(SketchTool.java:270)
	at sketch.SketchTool.loadSketches_MT(SketchTool.java:260)
	at sketch.SendSketch.processRemote(SendSketch.java:313)
	at sketch.SendSketch.process(SendSketch.java:306)
	at sketch.SendSketch.main(SendSketch.java:54)

Note that this error was thrown for a different sample than gave the error before ValueError => IndexError. This "empty" sample have been assembled correctly as it shows the correct assemblystats-output. I will have to look into this more tomorrow but will leave it here if anyone has some input before then. 
This script certainly is quite a puzzle...

**Comment by boulund** [comment]
What does the input files for the process that failed look like in the work directory? Could this be caused by a file system latency issue?

**Comment by emilio-r** [comment]
@boulund You're right that the /work is empty in this sample and only contains the .command-files, while the ones that passed have the .contigs etc there. I do remember now that this is something I and @abhi18av ran into before when diving into this issue, that some files do not get staged correctly between these steps.

**Comment by boulund** [comment]
I think it seems as if the changes in this PR are not responsible for the issues you observed during the testing @emilio-r , so I think we should merge this PR into develop, so we can test it together with other updates in the develop branch,

**Comment by emilio-r** [comment]
@boulund I was sceptical to merge since it contained known issues. But since we have shown that the issue is reproducible in master, and @thorellk showed that it exist when using other the other configs, I agree. I'll apply your suggested changes and then merge it.

**Pull request merged**